libraries/AP_HAL/UARTDriver.cpp

```diff
@@ -226,3 +232,128 @@ void AP_HAL::UARTDriver::log_stats(const uint8_t inst, StatsTracker &stats, cons
+        backend_loginfos = loginfo;
+    }
+    if (!log_thread_created) {
+        log_thread_created = true;
+        hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_HAL::UARTDriver::logging_start, void), "uart_log", 4096, AP_HAL::Scheduler::PRIORITY_IO, 0);
+    }
+
+    // write packet into buffer:
+    struct PACKED Header {
+        enum class Flag {
+            IS_WRITTEN_DATA = 1,
+        };
+        uint32_t magic = 0x7fe53b04U;
+        uint32_t time_ms;
+        uint16_t length;
+        uint8_t flags;
+    };
+
+    // note that we are assuming we can do this writing atomically.
+    // If there are multiple writers to the serial port that won't be
+    // the case.  We could reserve, write to and commit an iovec
+    // instead of this:
+
+    const auto space_required = sizeof(Header) + length;
+
+    if (loginfo->buf.space() < space_required) {
+        return;
+    }
+
+    Header header;
+    header.time_ms = AP_HAL::millis();
+    header.length = length;
+    if (is_written_data) {
+        header.flags |= uint8_t(Header::Flag::IS_WRITTEN_DATA);
+    }
+
+    loginfo->buf.write((uint8_t*)&header, sizeof(header));
+    loginfo->buf.write(data, length);
+}
+
+// a linked list of backends which are currently logging:
+AP_HAL::UARTDriver::LogInfo *AP_HAL::UARTDriver::backend_loginfos;
+bool AP_HAL::UARTDriver::log_thread_created;
+
+// logging loop, needs to be static to allow for re-alloc of GPS backends
+void AP_HAL::UARTDriver::logging_loop(void)
+{
+    while (true) {
+        hal.scheduler->delay(10);
+        for (auto *loginfo_ptr = backend_loginfos; loginfo_ptr != nullptr; loginfo_ptr = loginfo_ptr->next) {
+            auto &loginfo = *loginfo_ptr;
+            if (loginfo.fd == -2) {
+                // previously failed; ignore.
+                continue;
+            }
+            // check we have a log file open:
+            if (loginfo.fd == -1) {
+                char fname[] = "uartN_XXX.log";
+                fname[4] = '0' + loginfo.instance;
+                uint32_t lognum;
+                for (lognum=0; lognum<1000; lognum++) {
+                    struct stat st;
+                    hal.util->snprintf(&fname[6], 8, "%03u.log", (unsigned)lognum);
+                    if (AP::FS().stat(fname, &st) != 0) {
+                        break;
+                    }
+                }
+                loginfo.fd = AP::FS().open(fname, O_WRONLY|O_CREAT|O_APPEND);
+                if (loginfo.fd == -1) {
+                    // convert into a permanent failure:
+                    loginfo.fd = -2;
+                }
+            }
+            uint32_t n = 0;
+            const uint8_t *p;
+            while ((p = loginfo.buf.readptr(n)) != nullptr && n != 0) {
+                // short writes are unlikely and are ignored (only FS full errors)
+                AP::FS().write(loginfo.fd, p, n);
+                loginfo.buf.advance(n);
+                AP::FS().fsync(loginfo.fd);
```

Comment:
Fsync should be outside the loop. Also will generally be bad for performance but is probably necessary...